qiskit/algorithms/linear_solvers/hhl.py

```diff
@@ -0,0 +1,373 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The HHL algorithm."""
+
+from typing import Optional, Union, List, Callable, Tuple
+import numpy as np
+
+from qiskit.circuit import QuantumCircuit, QuantumRegister, AncillaRegister
+from qiskit.circuit.library import PhaseEstimation
+from qiskit.circuit.library.arithmetic.piecewise_chebyshev import PiecewiseChebyshev
+from qiskit.circuit.library.arithmetic.exact_reciprocal import ExactReciprocal
+from qiskit.opflow import Z, I, StateFn, TensoredOp, ExpectationBase, CircuitSampler, ListOp
+from qiskit.providers import Backend
+from qiskit.quantum_info import Operator
+from qiskit.quantum_info.operators.base_operator import BaseOperator
+from qiskit.utils import QuantumInstance
+
+from .linear_solver import LinearSolver, LinearSolverResult
+from .matrices.linear_system_matrix import LinearSystemMatrix
+from .observables.linear_system_observable import LinearSystemObservable
+
+
+class HHL(LinearSolver):
+    """The HHL algorithm to solve systems of linear equations"""
+
+    def __init__(self,
+                 epsilon: float = 1e-2,
+                 expectation: Optional[ExpectationBase] = None,
+                 quantum_instance: Optional[Union[Backend, QuantumInstance]] = None) -> None:
+        r"""
+        Args:
+            epsilon: Error tolerance of the approximation to the solution, i.e. if x is the exact
+                solution and ::math::`\tilde{x}` the one calculated by the algorithm, then
+                :math:`||x - \tilde{x}|| < epsilon`.
+
+
+        References:
+
+        [1]: Carrera Vazquez, A., Hiptmair, R., & Woerner, S. (2020).
+             Enhancing the Quantum Linear Systems Algorithm using Richardson Extrapolation.
+             `arXiv:2009.04484 <http://arxiv.org/abs/2009.04484>`_
+
+        """
+        super().__init__()
+
+        self._epsilon = epsilon
+        # Tolerance for the different parts of the algorithm as per [1]
+        self._epsilon_r = epsilon / 3  # conditioned rotation
+        self._epsilon_s = epsilon / 3  # state preparation
+        self._epsilon_a = epsilon / 6  # hamiltonian simulation
+
+        if quantum_instance is not None:
```

Comment:
If quantum_instance is None, I do not see any setter/getter like we normally have to set this later.

Reply:
added getter/setter

Reply:
Should this not create a new CircuitSampler if a new QuantumInstance is given. Also the quantum instance as given in the constructor is not stored so as to be able to be accessed via the getter

Reply:
Created a new CircuitSampler and stored the quantum instance.